File: sentimental-credit/credit.py
number = int(input("Number: "))
digits = list(str(number))

digits_reverse = []
for i in digits[::-1]:
    digits_reverse.append(i)

# Luhn's Algorithm
multiply = []
for i in range(1, len(digits), 2):
    multiply.append(str(2 * int(digits_reverse[i])))

for i in range(len(multiply)):
    if len(multiply[i]) == 2:
        multiply.append(multiply[i][0])
        multiply.append(multiply[i][1])
        multiply[i] = 0

sum = 0
for i in range(len(multiply)):
    sum += int(multiply[i])

not_multiply = []
for i in range(0, len(digits), 2):
    not_multiply.append(str(int(digits_reverse[i])))
for i in range(len(not_multiply)):
    sum += int(not_multiply[i])

# Check validity and type
second_digit_mastercard = ["1", "2", "3", "4", "5"]
if sum % 10 == 0:
    if len(digits) == 15 and (digits[0] == "3" and (digits[1] == "4" or digits[1] == "7")):
        print("AMEX")
    elif len(digits) == 16 and (digits[0] == "5" and digits[1] in second_digit_mastercard):
        print("MASTERCARD")
    elif len(digits) == 13 or len(digits) == 16 and digits[0] == "4":
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int find(long x, int y);

int main(void)
{
    // Prompt for input
    long n = get_long("Number: ");

    // Count the Luhn's Algorithm
    int sum = 0;
    int length = 0;
    long num = n;
    while (num > 0)
    {
        int digit = num % 10;
        if (length % 2 == 0)
        {
            sum += digit;
        }
        else
        {
            int doubled = digit * 2;
            sum += doubled / 10 + doubled % 10;
        }
        num /= 10;
        length++;
    }

    // Check starting digits and print result
    if (sum % 10 == 0)
    {
        if ((int) (n / (long) pow(10, length - 2)) == 34 ||
            (int) (n / (long) pow(10, length - 2)) == 37)
        {
            printf("AMEX\n");
        }
        else if ((int) (n / (long) pow(10, length - 2)) >= 51 &&
                 (int) (n / (long) pow(10, length - 2)) <= 55)
        {
            printf("MASTERCARD\n");
        }
        else if ((int) (n / (long) pow(10, length - 1)) == 4 ||
                 (int) (n / (long) pow(10, length - 2)) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int find(long x, int y)
{
    int digit = (x / (long) pow(10, y - 1)) % 10;
    return digit;
}
